### ISSUE TITLE
CSS syntax highlighting proposal

### ISSUE THREAD

**Issue by dragoncoder047** [w3c/csswg-drafts#9247]
This would be similar to the existing [CSS Painting API](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Painting_API).

Basically any text-color-affecting property (idk the proper term) would be able to use a new `syntax()` function to define which syntax grammar for the text that gets highlighted, and the individual tokens could be styled using the `::token()` pseudo-element.

So for example:

```css
::token(comment) { color: gray }
::token(tag-name) { color: red } /* these are really bad colors */
::token(attr-name) { color: blue }
::token(attr-value) { color: green }
textarea.language-html { color: syntax(html, black) }
 /* black is the fallback color if the text is not in a token */
```

and bam, instant syntax-highlited textarea.

Browsers would have built-in `html`, `css`, and `js` grammars, obviously (utilizing the existing parsers -- side effect is the grammar evolves with the language standard!), and for everything else there could be a plugin-based approach like the CSS Paint API. Something like `CSS.syntaxHighlighting.addModule("path/to/worker.js")`. Haven't bothered to think about how the worker would be passed the string to highlight (differential/incremental edits? highlight the whole string every time?), provide tokens to the browser (state machine? something else?), etc.

I don't think defining a *grammar* as opposed to a *worker* would be a good idea because that would be prone to bias towards a particular grammar definition, because there are tons of different kinds ([PrismJS](https://prismjs.com/), [highlight.js](https://highlightjs.org/), [Rainbow](https://craig.is/making/rainbows), [Pygments](https://pygments.org/), [GeSHi](http://qbnz.com/highlighter/), [Tree-sitter](https://tree-sitter.github.io/tree-sitter/), [TextMate](https://macromates.com/manual/en/language_grammars), even Emacs modes, just to name a few...)

One other thing I think would be useful would be a way for the CSS engine to signal it doesn't have a particular grammar registered, and prompt auto-loading via Javascript:

```js
CSS.syntaxHighlighting.addEventListener("language-not-found", e => {
    CSS.syntaxHighlighting.addModule(`/path/to/grammars/${e.language}.js`);
});
```

@LeaVerou would probably be a good person to comment on this proposal (for obvious reasons).

**Comment by LeaVerou** [comment]
I don't think this is not the appropriate level of abstraction for CSS. Do note that there is the [CSS Custom Highlight API](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Custom_Highlight_API), which is very similar to what you're describing here, but covers more broad use cases than syntax highlighting. In fact, we (Prism) are considering using once it gets Baseline browser support, though not sure we'd use it as a default.

Though do note that something like this does limit what you can do with syntax highlighting compared to the current approach. E.g. these Prism plugins would not be possible if we used the CSS Custom Highlight API:
- https://prismjs.com/plugins/autolinker/
- https://prismjs.com/plugins/inline-color/ (unless we allow generated content on `::highlight()`, or expand to more properties)
- https://prismjs.com/plugins/previewers/

Which might be an acceptable tradeoff, but something to have in mind.

**Comment by dragoncoder047** [comment]
> I don't think this is not the appropriate level of abstraction for CSS.

Do you support my proposal or not? Double-negatives are really confusing...

> * https://prismjs.com/plugins/inline-color/ (unless we allow generated content on `::highlight()`, or expand to more properties)
> * https://prismjs.com/plugins/previewers/

It might be possible to implement something like those using `::before` pseudo elements.

And honestly, I did expect there to be holes in this proposal with regards to interactive content (i.e. beyond simple `:active` or `:hover` styles).

One other use I did think of for this would be a textbook website could include some sort of worker that could apply token styles to all of the terms in the glossary, so they would appear bold (`::token(glossary) { font-weight: bold }`) and then the publisher doesn't have to rebuild the entire book when they update the glossary. Of course, actually liking to the glossary would not be possible with CSS alone.

**Comment by LeaVerou** [comment]
> Do you support my proposal or not? Double-negatives are really confusing...

Typo :) Fixed it now.

> One other use I did think of for this …

I think it would be useful to define your proposal in terms of delta from the Custom Highlight API. What does it offer that is better, and could these use cases be covered by an extension to this existing API, rather than a whole new one.

**Comment by dragoncoder047** [comment]
> I think it would be useful to define your proposal in terms of delta from the Custom Highlight API. What does it offer that is better, and could these use cases be covered by an extension to this existing API, rather than a whole new one.

The biggest thing is that the CSS Custom Highlight API doesn't work on textareas. See [this jsfiddle](https://jsfiddle.net/1bcsLe52/1/) -- the code tries to highlight the "print" in both a textarea and a pre, and it only works on the pre.

```html
<textarea cols=40>if True: print("hello")</textarea>
<pre>if False: print("bye")</pre>
```
```js
var area = document.querySelector("textarea");
var pre = document.querySelector("pre");
var areatext = area.childNodes[0];
var pretext = pre.childNodes[0];
var print1 = new Range();
print1.setStart(areatext, 9);
print1.setEnd(areatext, 14);
var print2 = new Range();
print2.setStart(pretext, 10);
print2.setEnd(pretext, 15);
CSS.highlights.set('function', new Highlight(print1, print2));
console.log("textarea has child node: ", areatext.toString());
console.log("pre has child node: ", pretext.toString());
```
```css
::highlight(function) { color: red }
```

The console.log()'s at the end both print `[object Text]` at the end for the `childNodes[0]` of both the textarea and the pre, and the range doesn't balk that it wasn't passed a valid node, but the highlight doesn't show up.

So we need some way to highlight text in textareas. Seriously, all of the major code editors out there only use a (hidden) textarea for capturing input events, and they re-implement all of the hard work of rendering the text onscreen in javascript.

**Comment by tabatkins** [comment]
Then yeah, I suspect pushing on *that* part (being able to style ranges in a textarea and other inputs) might be the place to start, before we step to a higher level and build in some parsing libraries. Outside of textareas, it looks like Custom Highlights does indeed solve the issue decently well, at least at a quick glance.

**Comment by LeaVerou** [comment]
Yeah, that sounds like a bug with the Highlight API that we should fix; definitely not a reason to design a whole other API.

**Comment by schenney-chromium** [comment]
It's not obvious if it's a browser bug or an API bug. If you can create a range you should be able to set a highlight on it, as I understand the spec. Does anyone happen to know if you can create the desired range for the textarea?

As reference, we can certainly draw spelling and grammar markers in textarea.

**Comment by dragoncoder047** [comment]
> As reference, we can certainly draw spelling and grammar markers in textarea.

You can?? I changed the CSS in the above jsfiddle to this:

```css
::highlight(function) { text-decoration-line: underline;
  text-decoration-style: wavy;
  text-decoration-color: red; }
```

and it drew the red squiggle only on the pre, but **NOT** on the textarea.

**Comment by bramus** [comment]
I believe this issue may be closed.

The original ask to do CSS syntax highlighting can be done with CSS parser (like Prism) + CSS Custom Highlights. See https://www.bram.us/2024/02/18/custom-highlight-api-for-syntax-highlighting/ for a write-up.

Using Custom Highlights on `<textarea>`s is tracked in https://github.com/w3c/csswg-drafts/issues/4603. A workaround you can use here, is to use `[contenteditable]`, as also described in https://www.bram.us/2024/02/18/custom-highlight-api-for-syntax-highlighting/#highlighting-contenteditable.

Hereby closing. Feel free to comment if you think this should be reopened.
